### DISCUSSION TITLE
Integration with django-polymorphic?

### DISCUSSION THREAD

**Discussion by furious-luke** [vitalik/django-ninja#365] [Q&A] [answered]
Hi there!

I'm in the process of experimenting with some options for creating a Django based API, and I'm wondering if Django Ninja has any kind of integration with Django Polymorphic? I've used Graphene (and Graphene Django) for a GraphQL based API in the past, and that integrates well with polymorphic models via the GraphQL "interface" standard. I can see that OpenAPI supports composed types and "oneOf" and "anyOf" keywords; can Django Ninja handle this already? And if not, are there plans to do so?

Thanks in advance!

**Comment by vitalik** [comment]
@furious-luke 

Not sure I completely understand your goals but `oneOf`, `anyOf` can be achieved with `Union` annotation

```Python
@api.post('/foo')
def some_foo(request, payload: Union[Foo, Bar, Test]):
    ...
```

**Reply by furious-luke** [reply]
Thanks for the reply @vitalik! Perhaps I can elaborate a little with an example. With Django Polymorphic we can define polymorphic models like this:

```python
class Person(Polymorphic):
    name = models.CharField()

class Manager(Person):
    role = models.CharField()

class Employee(Person):
    manager = models.ForeignKey(Manager)

# Returns all people named "John", including both managers and employees.
Person.objects.filter(name="John")
```

And so what I was hoping might be possible was to define a list endpoint like this:

```python
class PersonSchema(ModelSchema):
    class Config:
        model = Person
        model_fields = '__all__'

@api.get('/people', response=PersonSchema)
def list_people(request):
    return Person.objects.all()
```

And for Django Ninja to be able to know that there are multiple schema types in the set of people returned.

I can see how I might be able to achieve something similar using concrete types and union typings, however I suspect it will require knowledge of all concrete types in the system at the time of definition of the abstract type.

**Comment by vitalik** [comment, marked as answer]
@furious-luke 

OpenAPI forces you to describe **all** fields that you willing to output - so you would need to create union of all child models:

something like this:
```Python


class ManagerSchema(ModelSchema):
    class Config:
        model = Manager
        model_fields = '__all__'

class EmployeeSchema(ModelSchema):
    class Config:
        model = Employee
        model_fields = '__all__'


@api.get('/people', response=List[Union[ManagerSchema, EmployeeSchema]]) # !!!
def list_people(request):
    return Person.objects.all()

```

let me know if this works

**Reply by furious-luke** [reply]
Thanks again @vitalik, I appreciate your speedy replies!

Yeah no worries, that makes sense. The only issue remaining is that with GQL interfaces I don't need to define inherited schemas at the time of defining the base schema, which makes for a very convenient mechanism for cross app extensibility. This gives me a good starting point though, thanks!
